// angular2-image-zoom/angular2-image-zoom.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var image_zoom_directive_1 = require('./src/directives/image-zoom.directive');
exports.ImageZoom = image_zoom_directive_1.ImageZoom;
var image_zoom_lens_component_1 = require('./src/directives/image-zoom-lens.component');
exports.ImageZoomLens = image_zoom_lens_component_1.ImageZoomLens;
var image_zoom_container_component_1 = require('./src/directives/image-zoom-container.component');
exports.ImageZoomContainer = image_zoom_container_component_1.ImageZoomContainer;
var ImageZoomModule = (function () {
    function ImageZoomModule() {
    }
    ImageZoomModule = __decorate([
        core_1.NgModule({
            imports: [common_1.CommonModule],
            exports: [image_zoom_directive_1.ImageZoom],
            declarations: [image_zoom_directive_1.ImageZoom, image_zoom_container_component_1.ImageZoomContainer, image_zoom_lens_component_1.ImageZoomLens],
            entryComponents: [image_zoom_container_component_1.ImageZoomContainer, image_zoom_lens_component_1.ImageZoomLens]
        }), 
        __metadata('design:paramtypes', [])
    ], ImageZoomModule);
    return ImageZoomModule;
}());
exports.ImageZoomModule = ImageZoomModule;
